### PR TITLE
reimplement Compiler to handle nested CTEs in data tests

### PR DESCRIPTION
**Pull request by jtcohen6**
see #25

Here's the idea: You can override an adapter method within `dbt/adapters/sqlserver/impl.py` to change the `Compiler` that dbt will use when compiling models:
```python
    def get_compiler(self):
        from dbt.adapters.sqlserver import TSQLCompiler
        return TSQLCompiler(self.config)
```

Then, you can define that new compiler as a child class of dbt's core [`Compiler`](https://github.com/fishtown-analytics/dbt/blob/dev/kiyoshi-kuromiya/core/dbt/compilation.py#L143), in order to reimplement methods such as [`_add_ctes`](https://github.com/fishtown-analytics/dbt/blob/fec0e31a25d5b922cb1833cffcb5095eb4ee642b/core/dbt/compilation.py#L345). You could get as fancy as you want with this.

This code worked when I tested it in a local version of `dbt-postgres`, minus the `#` temp prefix

### CODE REVIEW

**Comment by alittlesliceoftom** [comment]
@jtcohen6 this seems like a promising method. Does the proposed change currently work to resolve #25 , or is it succesfully pointing at the new adapter, the new adapter isn't yet implementing the right logic?

**Comment by dataders** [comment]
my understanding is that we've been set up to solve the problem on our own with the approach that a temp table is made that is selected from rather than a CTE. our job is to modify the f-string, `compiled_node.compiled_sql` in the `__add_ctes()` method so that it does exactly that.

File: dbt/adapters/sqlserver/compilation.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+        # we need to wrap data test queries in such a way that we can 
+        # `select count(*)` at the end.
+        if isinstance(compiled_node, CompiledDataTestNode):
+            # create a temp table with the data test query, then get its count
+            name = self._get_dbt_test_name()
+            compiled_node.compiled_sql = f'''
+                create table {name} as \n{compiled_node.compiled_sql};\n
+                select count(*) from {name}
+            '''.strip()
```
Comment (dataders) [inline dbt/adapters/sqlserver/compilation.py:33]: this is the string in which we build a version of how data tests are executed.

**Comment by jtcohen6** [comment]
Yes, this taps into how dbt compiles data tests today (i.e. in v0.19.0-b1, the PR code maps to what's in `dev/kiyoshi-kuromiya`), and it's meant to demonstrate that you can override any behaviors of the builtin compiler. The way dbt compiles ephemeral models is a bit trickier—recursively prepending CTEs—but I bet you can find a way to rewrite that, too, so that it works in T-SQL.

FYI: my current plan is for the following dbt release (v0.20.0) to focus on tests, and I would _really like_ to take the data test wraparound code out of a python f-string (which is gross) and reimplement it in Jinja—at least as a macro, if not embedded within a test "materialization." If we manage to do that, overriding its behavior within the adapter would be a _lot_ more straightforward.

**Comment by alittlesliceoftom** [comment]
I guess recursively prepending ctes, should map to recursively chaining
them with possibly not that big a change

**Pull request closed without merging**